### PR TITLE
MRG, FIX: Test round-trip, fix EEGLAB ICA

### LINKED ISSUE #7657: The computation of the mixing matrix for ICA imported from EEGLAB ICA is broken
#### Describe the bug

Commit https://github.com/mne-tools/mne-python/pull/6938/commits/23c6aa9ac85411afba3dd3687787b1c112226546 introduced a bug in the PR https://github.com/mne-tools/mne-python/pull/6938 as described in the comment https://github.com/mne-tools/mne-python/pull/6938/commits/23c6aa9ac85411afba3dd3687787b1c112226546#r413113709

#### Steps to reproduce

Using an ICA with a number of components smaller than the number of channels, I get the following components with the code as patched with this commit:

image

and using seaborn clustermap on the correlation matrix between these components, I get:

image

#### Expected results

Now, if I revert this change, I get something more reasonable:

image

image

I unfortunately don't have the "gold truth" to compare with for the components at this time, but it already seems much more reasonable to me...

#### Proposal

I spotted that the issue here is the change from 

```
    ica.mixing_matrix_ = np.linalg.pinv(ica.pca_components_) @ eeg.icawinv
```
to
```
    ica._update_mixing_matrix()  # ==> self.mixing_matrix_ = linalg.pinv(self.unmixing_matrix_)
```

(Also, shouldn't "u * s" be "u @ s"? It does not seem to make any difference in my specific use case, but conceptually, it should be matrix multiplication, not a broadcasting.)

Apparently, the way EEGLAB defined its variable `linalg.pinv(self.unmixing_matrix_) != np.linalg.pinv(ica.pca_components_) @ eeg.icawinv`. I find it very hard to align the interfaces of MNE and EEGLAB because the approaches to ICA are a bit different, the naming is different, and so on, which makes it quite confusing. For example, how `u @ s` relates exactly to EEGLAB `icawinv` is unclear to me at this point. 

This seems to work:
```
    if n_components < n_ch:
        u, s, v = _safe_svd(eeg.icaweights, full_matrices=False)
        ica.unmixing_matrix_ = u @ s
        ica.pca_components_ = v
    else:
        [...]
    ica.mixing_matrix_ = np.linalg.pinv(ica.pca_components_) @ eeg.icawinv
```
The down side is that calling `ica._update_mixing_matrix()` at a later point would introduce a bug. So, an additional (ugly) fix is to store somewhere the value of `eeg.icawinv` and make  `ica._update_mixing_matrix()` of `ica.method == "imported_eeglab" `, which would looks like:

```
def _update_mixing_matrix(self):
    if self.method == "imported_eeglab":
        self.mixing_matrix_ = np.linalg.pinv(self.pca_components_) @ self._eeglab_icawinv         
    else:
        self.mixing_matrix_ = linalg.pinv(self.unmixing_matrix_)

def read_ica_eeglab(fname):
    [...]
    if n_components < n_ch:
        u, s, v = _safe_svd(eeg.icaweights, full_matrices=False)
        ica.unmixing_matrix_ = u @ s
        ica.pca_components_ = v
    else:
        [...]
    ica._eeglab_icawinv = eeg.icawinv  # Not great to add this attribute...
    ica._update_mixing_matrix()
```

FYI, simply doing
```
    if n_components < n_ch:
        u, s, v = _safe_svd(eeg.icaweights, full_matrices=False)
        # replace `ica.unmixing_matrix_ = u @ s` by
        ica.unmixing_matrix_ = np.linalg.pinv(np.linalg.pinv(ica.pca_components_) @ eeg.icawinv)
        ica.pca_components_ = v
    else:
        [...]
    ica._update_mixing_matrix() 
```
does not work. Don't ask me to explain why... Anyway, I can submit a PR with the proposed ugly changes and we can see if we can prettify it somehow... although, at this point, I am at a lost for a clean alignment of MNE and EEGLAB ICA interfaces, so I am not sure I can improve much on this.  In any case, I would argue that a working ugly code is preferable to an elegant broken one...

### LINKED ISSUE #8325: BUG?: ICA can blow up components
@hoechenberger not 100% sure this is related to #8313 and #8316 since those seem to have to do with ICA taking too long rather than doing something totally wrong. To me this code should work / be a no-op:
```
import numpy as np
import mne

data = np.random.RandomState(0).randn(10, 100000)
info = mne.create_info(10, 1000., 'eeg')
raw = mne.io.RawArray(data, info)
raw.set_eeg_reference()
assert np.linalg.matrix_rank(raw.get_data()) == 9

ica = mne.preprocessing.ICA(
    max_pca_components=10, n_pca_components=10, max_iter=1, verbose=True)
raw_new = ica.fit(raw).apply(raw.copy())
np.testing.assert_allclose(raw.get_data(), raw_new.get_data())
```
gives an `AssertionError`:
```
Mismatched elements: 1000000 / 1000000 (100%)
Max absolute difference: 4.53435389
Max relative difference: 70047.35032413
 x: array([[ 1.481445,  0.391694,  1.081697, ...,  0.370815, -1.556122,
        -1.275782],
       [-0.766405,  1.279594, -0.02692 , ...,  0.312448, -1.409486,...
 y: array([[ 6.455399e-05,  6.455399e-05,  6.455399e-05, ...,  6.455399e-05,
         6.455399e-05,  6.455399e-05],
       [ 3.581095e-03,  3.581095e-03,  3.581095e-03, ...,  3.581095e-03,...
```
It seems like this should work even if you get the number of components wrong. It suggests that something is wrong with the mixing/unmixing step that when the number of components equals the number of channels you don't get an identity transform here.

It also fails for `max_pca_components=10, n_pca_components=9` but passes for any value of `max_pca_components` (8, 9, 10) so long as `max_pca_components=9`.

If I remove this call:
```
        self.unmixing_matrix_ /= np.sqrt(
            self.pca_explained_variance_[sel])[None, :]  # whitening
```
Then things work okay.

### PR DESCRIPTION
**Pull request by larsoner**
Closes #8325 
Closes #8308
Closes #7657

### CODE REVIEW

**Comment by larsoner** [comment]
On this branch I get the same plots as in #8308:

Screenshot from 2020-10-01 14-18-38

@christian-oreilly @marcpabst can you try it on your datasets to see if it makes sense? I was able to unify the code so that no branching is required, which is nice.

File: mne/preprocessing/tests/test_ica.py
Comment (larsoner) [inline mne/preprocessing/tests/test_ica.py:156]: @hoechenberger this should pass for `n_pca_components=8` and `9` (and eventually probably error for `10` but it fails currently. This suggests that the `max_pca_components=0.9999` component picking does not work right.

File: mne/preprocessing/tests/test_ica.py
Comment (larsoner) [inline mne/preprocessing/tests/test_ica.py:163]: With 10 channels and an avg ref proj this gives 8 components instead of 9; if we use 1. instead, it works sometimes and not others because of numerical precision issues in the PCA.

If we want this to work (and I think we do) then we should change the logic to be "greater than or equal to" the given variance instead of "equal to". I consider this a bugfix. See previous [discussion](https://github.com/mne-tools/mne-python/pull/8321#discussion_r498159641).

Comment (hoechenberger) [reply mne/preprocessing/tests/test_ica.py:163]: Our current logic is "less than or equal to" :)

Comment (hoechenberger) [reply mne/preprocessing/tests/test_ica.py:163]: But I agree we should change it. That must be the reason why scikit-learn does it that way too.

**Comment by larsoner** [comment]
Okay this one is ready for review @agramfort @hoechenberger . We should wait to hear from @christian-oreilly and @marcpabst before merge, though

File: mne/utils/tests/test_numerics.py
Comment (hoechenberger) [inline mne/utils/tests/test_numerics.py:430]: I'm surprised this is `n_dim - 1`? I would have expected `n_dim` now, with the new adjustments?

**Review by hoechenberger — COMMENTED** [review]
Added a question, otherwise LGTM.
We need to do the same thing for `n_components`, for sake of consistency.

File: mne/utils/tests/test_numerics.py
Comment (larsoner) [reply mne/utils/tests/test_numerics.py:430]: This data is rank reduced, see a few lines above the comment:
```
X[:, -1] = np.mean(X[:, :-1], axis=-1)  # true X dim is ndim - 1
```

**Comment by larsoner** [comment]
Argh, I went to update the documentation, now I'm confused why we need `n_components`, `n_pca_components`, and `max_pca_components`. I thought we'd only need two: that the last one determined signal reconstruction, but now it says `n_pca_components` does this. I'll revert my doc changes (other than the refs) and hopefully we can fix it later.

**Comment by larsoner** [comment]
@hoechenberger I'm beginning to wonder if we need `max_pca_componetns` to be float. Really it should be enough to allow `n_components` to be a float I think. The instability of the inversion comes from dividing by `np.sqrt(self.pca_explained_variance_[:self.n_components_])`. So if `n_components` is set appropriately, it seems like there shouldn't be a problem.

**Comment by hoechenberger** [comment]
@larsoner
> @hoechenberger I'm beginning to wonder if we need `max_pca_componetns` to be float. Really it should be enough to allow `n_components` to be a float I think. The instability of the inversion comes from dividing by `np.sqrt(self.pca_explained_variance_[:self.n_components_])`. So if `n_components` is set appropriately, it seems like there shouldn't be a problem.

All `max_pca_components` are submitted to ICA for fitting; of those, only `n_components` are retained.

If it is like you say and we can get rid of the instability of choosing inappropriate `max_pca_components` and produce sane results, then things should be okay… Still I'm a little worried. Also, `max_pca_components` being a float could be just another way to limit how many components to fit ICA on – would be consistent with `n_components` and `n_pca_components`

**Comment by hoechenberger** [comment]
> @hoechenberger I'm beginning to wonder if we need `max_pca_componetns` to be float. Really it should be enough to allow `n_components` to be a float I think. The instability of the inversion comes from dividing by `np.sqrt(self.pca_explained_variance_[:self.n_components_])`. So if `n_components` is set appropriately, it seems like there shouldn't be a problem.

Yet another example of our API being way too complicated. We've had so many issues with this…

**Comment by hoechenberger** [comment]
> Argh, I went to update the documentation, now I'm confused why we need `n_components`, `n_pca_components`, and `max_pca_components`. I thought we'd only need two (`n_pca_components` and `max_pca_components`, and that `max_pca_components` determined signal reconstruction and `n_pca_components` was what was passed to the ICA algorithm, but this is not at all what our docs say. I'll revert my doc changes (other than the refs) and hopefully we can fix it later.

Yet another example of our API being way too complicated. We've had so many issues with this…

**Comment by larsoner** [comment]
Okay let me see if I can write the docs properly

**Comment by hoechenberger** [comment]
Maybe to clarify, my understanding is:

- max_pca_components: PCA components to keep
- n_components: number of ICA components that are retained after fitting ICA with max_pca_components --> the docs say otherwise, so either the docs are wrong or I made a mistake reading the sources
- n_pca_components: number of components (ICA + PCA!!!!) to use for signal reconstruction -> maybe this kwarg is just poorly named

**Comment by hoechenberger** [comment]
(edited my above post a couple times)

**Comment by larsoner** [comment]
> n_components: number of ICA components that are retained after fitting ICA on data with dimensionality reduced to max_pca_components 

The code has this:
```
        sel = slice(0, self.n_components_)
        if self.method == 'fastica':
            from sklearn.decomposition import FastICA
            ica = FastICA(whiten=False, random_state=random_state,
                          **self.fit_params)
            ica.fit(data_transformed[:, sel])
```
So I think the docs are correct -- ICA is fitted on the first `n_components` of the PCA.

`n_pca_components` isn't even used until `apply` so it only has to do with the apply step. I'll add this to the docs.

**Comment by larsoner** [comment]
Now I wonder if using `n_components` as float would have been enough and `max_pca_components` as float is overkill... I'll push a commit that shows setting `n_components` to a float also gives the behavior we want, even with `max_pca_components=n_chan`

**Comment by hoechenberger** [comment]
> So I think the docs are correct -- ICA is fitted on the first `n_components` of the PCA.

Oh ok, I got this mixed up then.

**Review by hoechenberger — APPROVED** [review]
LGTM once CI is green

File: mne/preprocessing/ica.py
Comment (hoechenberger) [inline mne/preprocessing/ica.py:285]:
```suggestion
    :meth:`apply`. These parameters can be used for dimensionality reduction of
```

**Comment by hoechenberger** [comment]
Thanks @larsoner! I'll want to try this on some of our data now before we move forward with the API redesign, ok?

**Pull request merged**
